File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/device/media/StatusCallback.java
package app.zxtune.device.media;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Handler;
import android.os.SystemClock;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.text.TextUtils;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;

import app.zxtune.Log;
import app.zxtune.R;
import app.zxtune.Releaseable;
import app.zxtune.TimeStamp;
import app.zxtune.core.Identifier;
import app.zxtune.core.ModuleAttributes;
import app.zxtune.fs.Vfs;
import app.zxtune.fs.VfsExtensions;
import app.zxtune.fs.VfsExtensionsKt;
import app.zxtune.fs.VfsFile;
import app.zxtune.fs.VfsObject;
import app.zxtune.fs.provider.VfsProviderClient;
import app.zxtune.playback.Callback;
import app.zxtune.playback.Item;
import app.zxtune.playback.PlaybackControl;
import app.zxtune.playback.PlaybackService;

//! Events gate from local service to mediasession
class StatusCallback implements Callback {

  private static final String TAG = StatusCallback.class.getName();

  private final Context ctx;
  private final MediaSessionCompat session;
  private final PlaybackStateCompat.Builder builder;
  private final Handler handler = new Handler();

  static Releaseable subscribe(Context ctx, PlaybackService svc, MediaSessionCompat session) {
    final StatusCallback cb = new StatusCallback(ctx, session);
    final PlaybackControl ctrl = svc.getPlaybackControl();
    //TODO: rework repeat/shuffle model
    session.setShuffleMode(ctrl.getSequenceMode().ordinal());
    session.setRepeatMode(ctrl.getTrackMode().ordinal());
    return svc.subscribe(cb);
  }

  private StatusCallback(Context ctx, MediaSessionCompat session) {
    this.ctx = ctx;
    this.session = session;
    this.builder = new PlaybackStateCompat.Builder();
    this.builder.setActions(
        PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS | PlaybackStateCompat.ACTION_PLAY_PAUSE |
            PlaybackStateCompat.ACTION_PLAY | PlaybackStateCompat.ACTION_PAUSE |
            PlaybackStateCompat.ACTION_STOP | PlaybackStateCompat.ACTION_SKIP_TO_NEXT
    );
  }

  @Override
  public void onInitialState(PlaybackControl.State state) {
    sendState(state, TimeStamp.EMPTY);
  }

  @Override
  public void onStateChanged(PlaybackControl.State state, TimeStamp pos) {
    sendState(state, pos);
  }

  private void sendState(PlaybackControl.State state, TimeStamp pos) {
    final boolean isPlaying = state == PlaybackControl.State.PLAYING;
    final int stateCompat = isPlaying
        ? PlaybackStateCompat.STATE_PLAYING
        : PlaybackStateCompat.STATE_STOPPED;
    final long position = pos.toMilliseconds();
    final float speed = isPlaying ? 1 : 0;
    builder.setState(stateCompat, position, speed, SystemClock.elapsedRealtime());
    session.setPlaybackState(builder.build());
    session.setActive(isPlaying);
  }

  @Override
  public void onItemChanged(Item item) {
    try {
      final Identifier dataId = item.getDataId();
      final MediaMetadataCompat.Builder builder = new MediaMetadataCompat.Builder();
      String title = item.getTitle();
      if (title.isEmpty()) {
        title = dataId.getDisplayFilename();
      }
      builder.putString(MediaMetadataCompat.METADATA_KEY_DISPLAY_TITLE, title);
      builder.putString(MediaMetadataCompat.METADATA_KEY_TITLE, title);
      final String author = item.getAuthor();
      if (!author.isEmpty()) {
        builder.putString(MediaMetadataCompat.METADATA_KEY_DISPLAY_SUBTITLE, author);
        builder.putString(MediaMetadataCompat.METADATA_KEY_ARTIST, author);
      } else {
        // Do not localize
        builder.putString(MediaMetadataCompat.METADATA_KEY_ARTIST, "Unknown artist");
      }
      putString(builder, ModuleAttributes.COMMENT, item.getComment());
      putString(builder, ModuleAttributes.PROGRAM, item.getProgram());
      putString(builder, ModuleAttributes.STRINGS, item.getStrings());
      builder.putLong(MediaMetadataCompat.METADATA_KEY_DURATION, item.getDuration().toMilliseconds());
      builder.putBitmap(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON, getLocationIcon(dataId.getDataLocation()));
      builder.putString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI, dataId.toString());
      builder.putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, item.getId().toString());
      fillObjectUrls(dataId.getDataLocation(), dataId.getSubPath(), builder);
      session.setMetadata(builder.build());
    } catch (Exception e) {
      Log.w(TAG, e, "onItemChanged()");
    }
  }

  @Override
  public void onError(final String e) {
    handler.post(() -> Toast.makeText(ctx, e, Toast.LENGTH_SHORT).show());
  }

  private static void putString(MediaMetadataCompat.Builder builder, String key,
                                @Nullable String value) {
    if (!TextUtils.isEmpty(value)) {
      builder.putString(key, value);
    }
  }

  @Nullable
  private Bitmap getLocationIcon(Uri location) {
    try {
      final Resources resources = ctx.getResources();
      final int id = getLocationIconResource(location);
      final Drawable drawable = ResourcesCompat.getDrawableForDensity(resources, id, 320/*XHDPI*/, null);
      if (drawable instanceof BitmapDrawable) {
        return ((BitmapDrawable) drawable).getBitmap();
      } else if (drawable != null) {
        final Bitmap result = Bitmap.createBitmap(64, 64, Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(result);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return result;
      }
    } catch (Exception e) {
      Log.w(TAG, e, "Failed to get location icon");
    }
    return null;
  }

  private int getLocationIconResource(Uri location) {
    try {
      final Uri rootLocation = new Uri.Builder().scheme(location.getScheme()).build();
      final VfsObject root = Vfs.resolve(rootLocation);
      Integer icon = VfsExtensionsKt.getIcon(root);
      if (icon != null) {
        return icon;
      }
    } catch (Exception e) {
      Log.w(TAG, e, "Failed to get location icon resource");
    }
    return R.drawable.ic_launcher;
  }

  private static void fillObjectUrls(Uri location,
                                     String subpath, MediaMetadataCompat.Builder builder) {
    try {
      final VfsObject obj = Vfs.resolve(location);
      putString(builder, VfsExtensions.SHARE_URL, (String) obj.getExtension(VfsExtensions.SHARE_URL));
      if (subpath.isEmpty() && obj instanceof VfsFile) {
        if (null != Vfs.getCacheOrFile((VfsFile) obj)) {
          putString(builder, VfsExtensions.FILE, VfsProviderClient.getFileUriFor(location).toString());
        }
      }
    } catch (Exception e) {
      Log.w(TAG, e, "Failed to get object urls");
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/device/media/StatusCallback.kt
package app.zxtune.device.media

import android.content.Context
import android.net.Uri
import android.os.Build
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.text.TextUtils
import app.zxtune.Logger
import app.zxtune.Releaseable
import app.zxtune.TimeStamp
import app.zxtune.Util
import app.zxtune.core.Identifier
import app.zxtune.core.ModuleAttributes
import app.zxtune.coverart.BitmapLoader
import app.zxtune.coverart.CoverartProviderClient
import app.zxtune.fs.Vfs
import app.zxtune.fs.VfsExtensions
import app.zxtune.fs.shareUrl
import app.zxtune.playback.Item
import app.zxtune.playback.PlayableItem
import app.zxtune.playback.PlaybackControl
import app.zxtune.playback.PlaybackService
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicReference

//! Events gate from local service to mediasession
internal class StatusCallback private constructor(
    private val ctx: Context,
    svc: PlaybackService,
    private val session: MediaSessionCompat
) : Releaseable {
    private val builder = PlaybackStateCompat.Builder()
    private val scope = CoroutineScope(CoroutineName("SessionStatusCallback") + Dispatchers.IO)
    private val coverartClient = CoverartProviderClient(ctx)
    private val lockscreenImageUrl = AtomicReference<Uri>()
    private val bitmapLoader by lazy {
        BitmapLoader("lockscreen", ctx, maxSize = 5, maxImageSize = 320)
    }

    init {
        builder.setActions(
            PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or PlaybackStateCompat.ACTION_PLAY_PAUSE or PlaybackStateCompat.ACTION_PLAY or PlaybackStateCompat.ACTION_PAUSE or PlaybackStateCompat.ACTION_STOP or PlaybackStateCompat.ACTION_SKIP_TO_NEXT
        )
        svc.playbackControl.run {
            session.setShuffleMode(sequenceMode.toShuffleMode())
            session.setRepeatMode(trackMode.toRepeatMode())
        }
        svc.nowPlaying.filterNotNull().onEach { onItemChanged(it) }.launchIn(scope)
        svc.state.onEach { onStateChanged(it.first, it.second) }.launchIn(scope)
    }

    override fun release() {
        scope.cancel()
    }

    private fun onStateChanged(state: PlaybackControl.State, pos: TimeStamp) {
        builder.setState(state.toState(), pos.toMilliseconds(), 1f)
        session.run {
            setPlaybackState(builder.build())
            isActive = state !== PlaybackControl.State.STOPPED
        }
    }

    private fun onItemChanged(item: Item) = try {
        val dataId = item.dataId
        val builder = MediaMetadataCompat.Builder().apply {
            val title = Util.formatTrackTitle(item.title, dataId)
            putString(MediaMetadataCompat.METADATA_KEY_DISPLAY_TITLE, title)
            putString(MediaMetadataCompat.METADATA_KEY_TITLE, title)
            val author = item.author
            if (author.isNotEmpty()) {
                putString(MediaMetadataCompat.METADATA_KEY_DISPLAY_SUBTITLE, author)
                putString(MediaMetadataCompat.METADATA_KEY_ARTIST, author)
            } else {
                // Do not localize
                putString(MediaMetadataCompat.METADATA_KEY_ARTIST, "Unknown artist")
            }
            putNonEmptyString(ModuleAttributes.TITLE, item.title)
            putNonEmptyString(ModuleAttributes.AUTHOR, item.author)
            putNonEmptyString(ModuleAttributes.COMMENT, item.comment)
            putNonEmptyString(ModuleAttributes.PROGRAM, item.program)
            putNonEmptyString(ModuleAttributes.STRINGS, item.strings)
            if (item is PlayableItem) {
                putNonEmptyString(
                    ModuleAttributes.CHANNELS_NAMES,
                    item.module.getProperty(ModuleAttributes.CHANNELS_NAMES, "")
                )
            }
            putLong(MediaMetadataCompat.METADATA_KEY_DURATION, item.duration.toMilliseconds())
            putLong(ModuleAttributes.SIZE, item.size)
            putString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI, dataId.toString())
            putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, item.id.toString())
        }
        fillObjectUrls(dataId, builder).takeIf { Build.VERSION.SDK_INT <= 29 }?.let {
            // https://developer.android.com/media/legacy/mediasession#album_artwork
            fillAlbumArtwork(it, builder)
        }
        session.setMetadata(builder.build())
    } catch (e: Exception) {
        LOG.w(e) { "onItemChanged()" }
    }

    private fun fillObjectUrls(
        dataId: Identifier, builder: MediaMetadataCompat.Builder
    ): Uri? = with(builder) {
        try {
            val obj = Vfs.resolve(dataId.dataLocation)
            putNonEmptyString(VfsExtensions.SHARE_URL, obj.shareUrl)
            coverartClient.getMediaUris(dataId)?.let { uris ->
                var preferableResult: Uri? = null
                for (key in arrayOf(
                    MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON_URI,
                    MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI,
                    MediaMetadataCompat.METADATA_KEY_ART_URI
                )) {
                    uris.getParcelable<Uri>(key)?.let {
                        preferableResult = it
                        putString(key, it.toString())
                    }
                }
                return preferableResult
            }
        } catch (e: Exception) {
            LOG.w(e) { "Failed to get object urls" }
        }
        return null
    }

    private fun fillAlbumArtwork(uri: Uri, builder: MediaMetadataCompat.Builder) {
        bitmapLoader.getCached(uri)?.bitmap?.let {
            LOG.d { "Cached album art" }
            builder.putBitmap(MediaMetadataCompat.METADATA_KEY_ART, it)
            return
        }
        lockscreenImageUrl.set(uri)
        scope.launch {
            bitmapLoader.load(uri).bitmap?.let { bitmap ->
                if (lockscreenImageUrl.compareAndSet(uri, null)) {
                    LOG.d { "Fetched album art" }
                    builder.putBitmap(MediaMetadataCompat.METADATA_KEY_ART, bitmap)
                    session.setMetadata(builder.build())
                } else {
                    LOG.d { "Drop outdated album art retrieval" }
                }
            }
        }
    }

    companion object {
        private val LOG = Logger(StatusCallback::class.java.name)

        @JvmStatic
        fun subscribe(ctx: Context, svc: PlaybackService, session: MediaSessionCompat) =
            StatusCallback(ctx, svc, session)
    }
}

private fun MediaMetadataCompat.Builder.putNonEmptyString(key: String, value: String?) = apply {
    if (!TextUtils.isEmpty(value)) {
        putString(key, value)
    }
}
